### ISSUE TITLE
Hyperparameters optimization

### ISSUE THREAD

**Issue by enricoferrero** [rstudio/keras#100]
This is not really an issue, just a question:
Is there a recommended way to optimise hyperparameters?

I know the Python version of keras can be wrapped in scikit-learn for this purpose and that other standalone packages exist (hyperopt, hyperas).

What is the best way to do it in R?
Thanks!

**Comment by topepo** [comment]
I'm currently adding a few MLP models to `caret` for parameter tuning. Others (RNN etc) can be easily adapted.

**Comment by jjallaire** [comment]
I think we'll need to create a hyperparameter tuning package that parallels the functionality of hyperopt, etc. In the meantime there is some functionality in the tfruns package (https://github.com/rstudio/tfruns) that you can use for this. 

First, install the latest dev versions of the keras package:

```r
devtools::install_github("rstudio/keras")
```

This will in turn install the latest version of the tfruns package, which you can use for parameter turning as illustrated in the example below. Obviously we will want to have a higher level vocabulary for doing this, but this at least provides something reasonable to work with for now.

```r
#' mnist_mlp.R -- trains a simple deep NN on the MNIST dataset.
#'
#' Gets to 98.40% test accuracy after 20 epochs (there is *a lot* of margin for
#' parameter tuning). 2 seconds per epoch on a K520 GPU.
#'
#' To tune, add FLAGS to the training script (see below for example), then 
#' perform runs with appropriate permutations of flags using the 
#' tfruns::training_run function. For example:
#'
#' library(tfruns)
#' training_run("mnist_mlp.R", flags = c(dropout_1 = 0.2))
#' training_run("mnist_mlp.R", flags = c(dropout_1 = 0.3))
#'
#' Then to view the results use the tfruns::ls_runs() function, which returns 
#' a data frame with all flags, training and evaluation metrics, etc. For example:
#'
#' library(tfruns)
#' ls_runs(order = "eval_acc")
#' 
#'    type                   run_dir eval_acc eval_loss metric_loss metric_acc metric_val_loss
#'   <chr>                     <chr>    <dbl>     <dbl>       <dbl>      <dbl>           <dbl>
#'  1 local runs/2017-08-15T17-01-39Z   0.9825    0.0984      0.0544     0.9851          0.1056
#'  2 local runs/2017-08-15T15-57-06Z   0.9817    0.1022      0.0491     0.9869          0.1127
#'  3 local runs/2017-08-14T19-39-57Z   0.9816    0.0999      0.0497     0.9866          0.1147
#'  4 local runs/2017-08-15T16-51-55Z   0.9815    0.1111      0.0540     0.9859          0.1145
#'  5 local runs/2017-08-14T19-44-30Z   0.9814    0.1016      0.0510     0.9868          0.1055
#'  6 local runs/2017-08-14T16-54-01Z   0.9812    0.0998      0.0539     0.9858          0.1111
#'  7 local runs/2017-08-14T15-44-09Z   0.9810    0.1083      0.0518     0.9871          0.1183
#'  8 local runs/2017-08-14T18-18-01Z   0.9808    0.0992      0.0508     0.9860          0.1086
#'  9 local runs/2017-08-15T12-32-29Z   0.9804    0.0999      0.0490     0.9860          0.1139
#' 10 local runs/2017-08-14T16-52-36Z   0.9794    0.1105      0.0489     0.9866          0.1130
#' # ... with 11 more rows, and 14 more variables: metric_val_acc <dbl>, flag_learning_rate <dbl>,
#' #   flag_max_steps <int>, flag_data_dir <chr>, flag_fake_data <lgl>, samples <int>,
#' #   batch_size <int>, epochs <int>, epochs_completed <int>, start <dttm>, end <dttm>,
#' #   completed <lgl>

library(keras)

# flags for parameter tuning.
FLAGS <- flags(
  flag_numeric('dropout_1', 0.04, 'First dropout'),
  flag_numeric('dropout_2', 0.3, 'Second dropout'),
  flag_integer('dense_units', 128, 'Units in dense layer')
)

batch_size <- 128
num_classes <- 10
epochs <- 30

# the data, shuffled and split between train and test sets
mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

dim(x_train) <- c(nrow(x_train), 784)
dim(x_test) <- c(nrow(x_test), 784)

x_train <- x_train / 255
x_test <- x_test / 255

cat(nrow(x_train), 'train samples\n')
cat(nrow(x_test), 'test samples\n')

# convert class vectors to binary class matrices
y_train <- to_categorical(y_train, num_classes)
y_test <- to_categorical(y_test, num_classes)

model <- keras_model_sequential()
model %>% 
  layer_dense(units = 256, activation = 'relu', input_shape = c(784)) %>% 
  layer_dropout(rate = FLAGS$dropout_1) %>% 
  layer_dense(units = FLAGS$dense_units, activation = 'relu') %>%
  layer_dropout(rate = FLAGS$dropout_2) %>%
  layer_dense(units = 10, activation = 'softmax')

summary(model)

model %>% compile(
  loss = 'categorical_crossentropy',
  optimizer = optimizer_rmsprop(),
  metrics = c('accuracy')
)

history <- model %>% fit(
  x_train, y_train,
  batch_size = batch_size,
  epochs = epochs,
  verbose = 1,
  callbacks = callback_tensorboard(),
  validation_split = 0.2
)

plot(history)
  
score <- model %>% evaluate(
  x_test, y_test,
  verbose = 0
)
  
cat('Test loss:', score[[1]], '\n')
cat('Test accuracy:', score[[2]], '\n')
```

**Comment by enricoferrero** [comment]
@topepo Are you also looking into mlr by any chance?
They have a neat and elegant interface to incorporate learners.

https://mlr-org.github.io/
https://github.com/mlr-org/mlr

CC @jakob-r @berndbischl @larskotthoff

**Comment by enricoferrero** [comment]
@jjallaire Thanks, that looks pretty good already!

Can you also optimize activation function types, number of epochs, loss function, optimizers, etc with tfruns?

A standalone package would be neat, are you also considering piggybacking metapackages like caret and mlr for tuning? CC @berndbischl

**Comment by jjallaire** [comment]
Yes, the `FLAGS` mechanism can tune anything which you can manifest as a variable in your script. Since many Keras model attributes (optimizer, loss, etc.) can actually be specified via strings it will be very easy to tune these things. e.g.

```r
FLAGS <- flags(
  flag_numeric('loss', 'categorical_crossentropy', 'Loss function'),
  flag_numeric('optimizer', 'adam', Optimizer')
)

model %>% compile(
  loss = FLAGS$loss,
  optimizer = FLAGS$optimizer,
  metrics = c('accuracy')
)
```


As Max said there is already some stuff going into the next version of caret. Hopefully the `flags` / `training_run` mechanism is flexible enough that it can fit well into just about any existing tuning framework.

**Comment by topepo** [comment]
@enricoferrero 

> Are you also looking into `mlr` by any chance?

No, not really. I created `caret` and that's good enough for me =] More than one options is good though. I've used `rBayesianOptimization` with `keras` to tune and that works well too. 

I'm also putting together tidy modeling packages for parameter tuning (to complement `recipes`, `rsample`, etc). These will work with `keras` models too.

**Comment by enricoferrero** [comment]
@topepo oops, sorry, I didn't realise who I was talking to! :sweat_smile:

**Comment by KafeelBasha** [comment]
How can I tune 'batch_size' and 'dropout' parameters in Keras?

**Comment by jjallaire** [comment]
I recommend that you start with the tfruns package (https://tensorflow.rstudio.com/tools/tfruns/). See the docs on hyperparameter tuning here: https://tensorflow.rstudio.com/tools/tfruns/articles/tuning.html

For a more sophisticated approach you can also use CloudML: https://tensorflow.rstudio.com/tools/cloudml/articles/tuning.html

**Comment by KafeelBasha** [comment]
Thanks a lot.

**Comment by topepo** [comment]
There is some code (for SVMs) that can be adapted in [this repo](https://github.com/topepo/Optimization-Methods-for-Tuning-Predictive-Models).
